rdtools/degradation.py

```diff
@@ -74,82 +71,107 @@ def degradation_with_ols(normalized_energy):
+    Classical decomposition routine from Dirk Jordan, Chris Deline, and Michael Deceglie
 
     Parameters
     ----------
-    normalized_energy: Pandas Series (numeric) containing corrected performance ratio, 
-        with index being monthly frequency in timestamp format
-    interpolate_flag: boolean flag to either interpolate missing data or fill 
-        with the median value (default)
+    normalized_energy: Pandas Time Series (numeric)
+        Daily or lower frequency time series of normalized system ouput.
+        Must be regular time series.
 
     Returns
     -------
     degradation_values: dictionary
-        Contains values for annual degradation rate and standard error
-        'Rd_pct', 'slope', 'intercept', 'Rd_stderr_pct', 'dataframe'
-        where dataframe contains annual rolling mean values
+        Contains values for annual degradation rate as %/year ('Rd_pct'),
+        slope, intercept, root mean square error of regression ('rmse'),
+        standard error of the slope ('slope_stderr') and intercept ('intercept_stderr'),
+        the 68.2% confidence interval for Rd ('Rd_CI'),
+        least squares RegressionResults object ('ols_results'),
+        pandas series for the annual rolling mean ('series'),
+        Mann-Kendall test trend ('mk_test_trend')
     '''
     
     normalized_energy.name = 'normalized_energy'
-    dataframe = normalized_energy.to_frame()
-
-    # try to extract month information from the dataframe index. Fill missing data with median value
-    energy_median = normalized_energy.median()
+    df = normalized_energy.to_frame()
+    
+    df_check_freq = df.copy()
 
-    # check for DatetimeIndex
-    if isinstance(dataframe.index, pd.DatetimeIndex):
-        dataframe = dataframe.resample('MS').mean()
-        
-        if interpolate_flag:
-	    dataframe = dataframe.interpolate()
-        else:
-            # append the median value to missing months
-            dataframe = dataframe.fillna(value = energy_median)
+    df_check_freq = df_check_freq.dropna()
 
-    # assume an array of months
-    dataframe['Month'] = np.arange(0, len(dataframe))
+    if df_check_freq.index.freq is None:
+        raise ValueError('Classical decomposition requires a regular time series with'
+                         ' defined frequency and no missing data.')
 
-    if pd.infer_freq(dataframe.index) == 'MS' and len(normalized_energy) > 12:
-        dataframe = dataframe.rolling(window = 12, center = True).mean()
-        
-    dataframe = dataframe.dropna()
-    y2 = dataframe['normalized_energy']
-    x2 = dataframe['Month']
+    #calculate a years column as x value for regression, ignoreing leap years
+    day_diffs = (df.index - df.index[0])
+    df['days'] = day_diffs.astype('timedelta64[s]')/(60*60*24)
+    df['years'] = df.days/365.0
     
+    #Compute yearly rolling mean to isolate trend component using moving average
+    it = df.iterrows()
+    energy_ma = []
+    for i, row in it:
```

Comment:
Good question. We need a centered rolling mean for seasonal decomposition. However `center = True` is not implemented for  offset-based windows e.g. `.rolling('365D', center = True)` raises a `NotImplementedError`.